tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,763 @@
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DtDouble
+func (t *Tensor) AsFloat64() (res []float64, err error) {
+	if DtDouble != t.DataType() {
+		err = &ErrInvalidTensorType{
+			tensorType:   t.DataType(),
+			expectedType: DtDouble,
+		}
+		return
+	}
+
+	if t.DoubleVal != nil {
+		return t.DoubleVal, nil
+	}
+
+	data := t.Data()
+	res = make([]float64, len(data)/cBytesFloat64)
+	for i := range res {
+		res[i] = math.Float64frombits(binary.LittleEndian.Uint64(data[i*cBytesFloat64 : (i+1)*cBytesFloat64]))
+	}
+	t.DoubleVal = res
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return
+}
+
+// AsInt32 returns the content of the tensor as a slice of int32 if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DtUint8
+//  - DtInt8
+//  - DtInt16
+//  - DtInt32
+func (t *Tensor) AsInt32() (res []int32, err error) {
+	if t.IntVal != nil {
+		return t.IntVal, nil
+	}
+
+	data := t.Data()
+	switch t.DataType() {
+	case DtInt8, DtUint8:
+		res = make([]int32, len(data))
+		for i, v := range data {
+			res[i] = int32(v)
+		}
+	case DtInt16:
+		res = make([]int32, len(data)/cBytesUint16)
+		for i := range res {
+			res[i] = int32(binary.LittleEndian.Uint16(data[i*cBytesUint16 : (i+1)*cBytesUint16]))
+		}
+	case DtInt32:
+		res = make([]int32, len(data)/cBytesInt32)
+		for i := range res {
+			res[i] = int32(binary.LittleEndian.Uint32(data[i*cBytesInt32 : (i+1)*cBytesInt32]))
+		}
+	default:
+		err = &ErrInvalidTensorType{
+			tensorType:   t.DataType(),
+			expectedType: DtInt32,
+		}
+		return
+	}
+
+	t.IntVal = res
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return
+}
+
+// AsInt64 returns the content of the tensor as a slice of int64 if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DtInt64
+func (t *Tensor) AsInt64() (res []int64, err error) {
+	if DtInt64 != t.DataType() {
+		err = &ErrInvalidTensorType{
+			tensorType:   t.DataType(),
+			expectedType: DtInt64,
+		}
+		return
+	}
+
+	if t.Int64Val != nil {
+		return t.Int64Val, nil
+	}
+
+	data := t.Data()
+	res = make([]int64, len(data)/cBytesInt64)
+	for i := range res {
+		res[i] = int64(binary.LittleEndian.Uint64(data[i*cBytesInt64 : (i+1)*cBytesInt64]))
+	}
+	t.Int64Val = res
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return
+}
+
+// AsBool returns the content of the tensor as a slice of bool if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DtBool
+func (t *Tensor) AsBool() (res []bool, err error) {
+	if DtBool != t.DataType() {
+		err = &ErrInvalidTensorType{
+			tensorType:   t.DataType(),
+			expectedType: DtBool,
+		}
+		return
+	}
+
+	if t.BoolVal != nil {
+		return t.BoolVal, nil
+	}
+
+	data := t.Data()
+	res = make([]bool, len(data))
+	for i, v := range data {
+		res[i] = v == 1
```

Comment:
Could you parenthesize the logical comparison to make it clear?  res[i] = (v == 1)
a few more characters, but it makes it more obvious that it's not a weird typo. :)